src/Streamly/Internal/Data/Producer/Type.hs

```diff
@@ -0,0 +1,163 @@
+-- |
+-- Module      : Streamly.Internal.Data.Producer.Type
+-- Copyright   : (c) 2021 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- See "Streamly.Internal.Data.Producer" for introduction.
+--
+
+module Streamly.Internal.Data.Producer.Type
+    (
+    -- * Type
+    Producer (..)
+
+    -- * Producers
+    , nil
+    , nilM
+    , unfoldrM
+    , fromList
+
+    -- * Mapping
+    , translate
+    , lmap
+
+    -- * Nesting
+    , NestedLoop (..)
+    , concat
+    )
+where
+
+#include "inline.hs"
+
+import Fusion.Plugin.Types (Fuse(..))
+import Streamly.Internal.Data.Stream.StreamD.Step (Step(..))
+import Prelude hiding (concat)
+
+------------------------------------------------------------------------------
+-- Type
+------------------------------------------------------------------------------
+
+-- | A @Producer m a b@ is a generator of a stream of values of type @b@ from a
+-- seed of type 'a' in 'Monad' @m@.
+--
+-- /Internal/
+
+data Producer m a b =
+    -- | @Producer step inject extract@
+    forall s. Producer (s -> m (Step s b)) (a -> m s) (s -> m a)
+
+------------------------------------------------------------------------------
+-- Producers
+------------------------------------------------------------------------------
+
+{-# INLINE nilM #-}
+nilM :: Monad m => (a -> m c) -> Producer m a b
+nilM f = Producer step return return
+
+    where
+
+    {-# INLINE_LATE step #-}
+    step x = f x >> return Stop
+
+{-# INLINE nil #-}
+nil :: Monad m => Producer m a b
+nil = nilM (\_ -> return ())
+
+{-# INLINE unfoldrM #-}
+unfoldrM :: Monad m => (a -> m (Maybe (b, a))) -> Producer m a b
+unfoldrM next = Producer step return return
+  where
+    {-# INLINE_LATE step #-}
+    step st = do
+        r <- next st
+        return $ case r of
+            Just (x, s) -> Yield x s
+            Nothing -> Stop
+
+-- | Convert a list of pure values to a 'Stream'
+--
+-- /Internal/
+{-# INLINE_LATE fromList #-}
+fromList :: Monad m => Producer m [a] a
+fromList = Producer step return return
+
+    where
+
+    {-# INLINE_LATE step #-}
+    step (x:xs) = return $ Yield x xs
+    step [] = return Stop
+
+------------------------------------------------------------------------------
+-- Mapping
+------------------------------------------------------------------------------
+
+-- | Interconvert the producer between two interconvertible input types.
+--
+-- /Internal/
+{-# INLINE_NORMAL translate #-}
+translate :: Functor m =>
+    (a -> c) -> (c -> a) -> Producer m c b -> Producer m a b
+translate f g (Producer step inject extract) =
+    Producer step (inject . f) (fmap g . extract)
+
+-- | Map the producer input to another value of the same type.
+--
+-- /Internal/
+{-# INLINE_NORMAL lmap #-}
+lmap :: (a -> a) -> Producer m a b -> Producer m a b
```

Comment:
We can probably remove this or rename this. I don't think this justifies `lmap`.

Reply:
A search for `f :: a -> a -> g a -> g a` on hoogle gave only fmap/liftA/liftM. Its a map even though it is restricted to an endomorphism. Any alternative names? If we remove it can we achieve the same functionality in some other way?